cl/validator/attestation_producer/attestation_producer.go

```diff
@@ -56,7 +59,49 @@ func New(ctx context.Context, beaconCfg *clparams.BeaconChainConfig) Attestation
 	return p
 }
 
-func (ap *attestationProducer) ProduceAndCacheAttestationData(baseState *state.CachingBeaconState, slot uint64, committeeIndex uint64) (solid.AttestationData, error) {
+func (ap *attestationProducer) computeTargetCheckpoint(tx kv.Tx, baseState *state.CachingBeaconState, slot uint64) (solid.Checkpoint, error) {
+	baseStateBlockRoot, err := baseState.BlockRoot()
+	if err != nil {
+		return solid.Checkpoint{}, err
+	}
+
+	targetEpoch := slot / ap.beaconCfg.SlotsPerEpoch
+	epochStartTargetSlot := targetEpoch * ap.beaconCfg.SlotsPerEpoch
+	var targetRoot libcommon.Hash
+	if tx != nil {
+		targetRoot, err = beacon_indicies.ReadCanonicalBlockRoot(tx, epochStartTargetSlot)
+		if err != nil {
+			return solid.Checkpoint{}, fmt.Errorf("failed to get targetRoot at slot from db %d: %w", epochStartTargetSlot, err)
+		}
+		if targetRoot != (libcommon.Hash{}) {
+			return solid.Checkpoint{
+				Root:  targetRoot,
+				Epoch: targetEpoch,
+			}, nil
+		}
+	}
+
+	if epochStartTargetSlot >= baseState.Slot() {
+		targetRoot = baseStateBlockRoot
+	} else {
```

Comment:
Curious, is `beacon_indicies.ReadCanonicalBlockRoot` more precise (or correct) than read block by `baseState.GetBlockRootAtSlot`? 
Can we remove the `else` case here? Is it still useful or just for unittest purpose?

Reply:
It depends - I added the DB as a second layer of checking because it `GetBlockRootAtSlot` might not be avaiable in some annoying corner-case